### ISSUE TITLE
how to generate smpl boundry free image?

### ISSUE THREAD

**Issue by GuiZhaoyang** [zengwang430521/DecoMR#18]
@zengwang430521 I have seen the smpl boundry free image in the paper and I want to generate it. But I did not find the part that generates  smpl boundry free image in the code. Any help would be greatly appreciated.

**Comment by zengwang430521** [comment]
Hi, the generation of IUV image is included in the **_preprocess_datasets.py_**.
With the arguments --gt_iuv, this function will generate the GT IUV image of H36M and UP-3D datasets.

**Comment by GuiZhaoyang** [comment]
Hi, I have generated iuv image. But I wanna generate the transferred RGB image which similar to the Transferred Pixel in Figure 5 in your paper. Any help would be greatly appreciated.

**Comment by zengwang430521** [comment]
You can use the function _models.dense_cnn.warp_feature()_ with the RGB image and IUV image as the inputs. 
https://github.com/zengwang430521/DecoMR/blob/f96526ba9547d23cddd5931fa3fe75a49927e3f3/models/dense_cnn.py#L18

'dp_out' can be gotten from IUV image. The first channel is the foreground-background mask and the last 2 channels are the UV coordinates normalized in [0, 1]. You can fins more details from the codes.

**Comment by GuiZhaoyang** [comment]
@zengwang430521 Hi, I have use the function, but it seem that the warpped uv_map and image result are wrong. This is my code, the warp_feature function is not modified.

`   dataset = BaseDataset(options, dataset, use_augmentation=False, is_train=False, use_IUV=True)
    dataloader = DataLoader(dataset, batch_size=2, shuffle=False)
    item = next(iter(dataloader))
    img, iuv = item['img'], item['gt_iuv']
    dtype = iuv.dtype
    iuv[:, 0] = (iuv[:, 0].unsqueeze(1) > 0).type(dtype)
    iuv[:, 1:] = iuv[:, 1:] / 255.0
    gt_uv_shape = iuv[:, 1:]
    dp_out = torch.cat((gt_mask_shape, gt_uv_shape), 1)
    batch_size = img.shape[0]
    warped_feature = warp_feature(dp_out, img, img.shape[2])
    trans_uv, trans_img = warped_feature[:, :3], warped_feature[:, 3:]
    save_uv_map(item['imgname'], 'examples/BF_UV', trans_uv)
    save_Img(item['imgname'], 'examples/BF_Img', trans_img)
`
The function of save_uv_map and save_Img are as following:
`def save_uv_map(save_names, save_dir, uv_maps):
    if uv_maps.shape[3] != 3:
        uv_maps = uv_maps.permute(0, 2, 3, 1)
    save_uv_maps = uv_maps.cpu().numpy()
    for i in range(save_uv_maps.shape[0]):
        save_name = os.path.join(save_dir, save_names[i].split('/')[-1])
        cv2.imwrite(save_name, save_uv_maps[i] * 255)

def save_Img(save_names, save_dir, saved_img):
    saved_img = saved_img * torch.tensor([0.229, 0.224, 0.225], device=saved_img.device).reshape(1,3,1,1)
    saved_img = saved_img + torch.tensor([0.485, 0.456, 0.406], device=saved_img.device).reshape(1,3,1,1)
    saved_img = saved_img.permute(0, 2, 3, 1)

    saved_img = saved_img.cpu().numpy()
    for i in range(saved_img.shape[0]):
        save_name = os.path.join(save_dir, save_names[i].split('/')[-1])
        cv2.imwrite(save_name, saved_img[i, :, :, ::-1] * 255)`

**Comment by zengwang430521** [comment]
Hi, could you please provide the full _.py_ file you used.

**Comment by zengwang430521** [comment]
Hi, 
**item['img']** is the normalized image. If you want to get results for visualization, you may need to use **item['img_orig']**

**Comment by GuiZhaoyang** [comment]
Hi, I have use the item['img_orig'], but it still seems wrong.
![Image](https://i.loli.net/2021/02/22/koY9p7VTdBMz4ug.jpg)
![gt BF uv map use smpl parameters](https://i.loli.net/2021/02/22/slwL6vMoj5qGH2A.jpg)
![wrong BF uv map use iuv warpped](https://i.loli.net/2021/02/22/aE7kvu9egM5Yqnh.jpg)
![wrong BF img use iuv warpped](https://i.loli.net/2021/02/22/vl8Z43UW5ncbEYP.jpg)

**Comment by zengwang430521** [comment]
Hi, the color is a little strange. It seems that you may need to change the channel order when saving the warped image.

**Comment by GuiZhaoyang** [comment]
Hi, I acquire the warpped feature with 6 channels and I use the 0~2 channels as BF_uv and 3~5 channels as BF_img.

**Comment by zengwang430521** [comment]
Hi, for a input feature map with C channels, the output of warp_feature() is with C+3 channels.
The channel [0:C] is the warped feature. In this case, it's the warped pixels. And the channel [C] and [C+1] is the **_x_** and **_y_** coordinates of the warped pixels. The last channel is the mask to determine whether this uv location is with warped pixels or not.

**Comment by zengwang430521** [comment]
If you want to get the warped RGB pixels, using the first 3 channels is enough.

**Comment by GuiZhaoyang** [comment]
Hi, I have used the first 3 channels to generate the BF UV map which has shown above, but it is seemed still wrong.

**Comment by zengwang430521** [comment]
Could you please provide the python script you use? I want to run it to figure out the problem.

**Comment by GuiZhaoyang** [comment]
Ok, I have put the script in the link.
[https://github.com/GuiZhaoyang/my_code/blob/main/generate_UV.py](url)

**Comment by GuiZhaoyang** [comment]
And the realted script alse in the branch.

**Comment by zengwang430521** [comment]
Hi, I have sent a pull request in your project and you may try the modified codes.

**Comment by GuiZhaoyang** [comment]
It seems that I have acquire the right result. And thanks very much for your help.
